### PR TITLE
Add auction history get and list endpoints

### LINKED ISSUE #417: Loan Auctions API
#### What would you like to be added:

Exposing loan auctions API, to be added into `LoanController.ts` at server with `Loan.ts` whale-api-client.

```
GET: /loan/auctions
GET: /loan/auctions/:id
```

/triage accepted
/priority important-soon
/area module-api

### PR DESCRIPTION
**Pull request by jingyi2811**
What kind of PR is this?:
/kind feature

What this PR does / why we need it:
Add auction history get and list endpoints

Which issue(s) does this PR fixes?:
Fixes #417

Additional comments?:

### CODE REVIEW

**Comment by jingyi2811** [comment]
If C++ side has fixed the pagination issue, rerun the failed CI test will pass.

File: src/module.api/loan.auctionhistory.controller.e2e.ts [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+      })
+    }
+    )
+  })
+})
+
+it('should listAuctionHistory with pagination', async () => {
+  const first = await controller.listAuctionHistory({ size: 2 })
+  expect(first.data.length).toStrictEqual(2)
+  expect(first.page?.next).toStrictEqual(`${first.data[1].vaultId}${first.data[1].blockHeight}`)
```
Comment (ivan-zynesis) [inline src/module.api/loan.auctionhistory.controller.e2e.ts:390]: If I understand L375 correctly, each data is for 1 batch, there is possibility for missing batch?

eg actual history (let say descending ordered)
[
  <vaultA><1000>, // batch 0
  <vaultA><1000>, // batch 1
  <vaultA><1000> // batch 2
  <vaultA><999> // batch 0
]

page1 = `listAuctionHistory({ size: 2 })`
page2 = `listAuctionHistory({ size: 2, next: <vaultA><1000> })`
`<vaultA><1000><batch2>` will be excluded right?

@Jouzo @surangap

File: src/module.api/loan.controller.ts [hunk 2]
```diff
@@ -185,6 +186,17 @@ export class LoanController {
     return await this.vaultService.listAuction(query)
   }
 
+  /**
+   * Paginate loan auction history.
+   *
+   * @param {PaginationQuery} query
+   * @return {Promise<ApiPagedResponse<LoanAuctionHistory>>}
+   */
+  @Get('/auctionhistory')
```
Comment (canonbrother) [inline src/module.api/loan.controller.ts:195]:
```suggestion
  @Get('/auctions/history')
```

File: src/module.api/loan.vault.service.ts [hunk 3]
```diff
@@ -196,6 +225,19 @@ export class LoanVaultService {
       interestRate: scheme.interestrate.toFixed()
     }
   }
+
+  private mapLoanAuctionHistory (detail: ListAuctionHistoryDetail): LoanAuctionHistory {
```
Comment (canonbrother) [inline src/module.api/loan.vault.service.ts:229]: it doesn't need to map if `ListAuctionHistoryDetail` and `LoanAuctionHistory` is same..

**Review by canonbrother — COMMENTED** [review]
file naming -> use `.` as delimiter
eg:
loan.auction.history.test.ts

**Comment by fuxingloh** [comment]
> file naming -> use `.` as delimiter eg: loan.auction.history.test.ts

@canonbrother need your help with taking over this PR

**Comment by fuxingloh** [comment]
I am not sure what's the issue but if we can correct the problem on whale side to get the correct behaviour we should just do it at this stage.

**Comment by canonbrother** [comment]
> I am not sure what's the issue but if we can correct the problem on whale side to get the correct behaviour we should just do it at this stage.

ahh.. ya.. you're right.. its whale issue

File: src/module.api/loan.vault.service.ts [hunk 4]
```diff
@@ -100,6 +101,37 @@ export class LoanVaultService {
...
+          const full = await this.client.loan.listAuctionHistory('all', { limit: limit + 30, maxBlockHeight })
+          const index = full.findIndex(f => f.vaultId === vaultId)
+          if (index > 0) {
+            const sliced = full.slice(index + 1, index + 1 + query.size)
+            return ApiPagedResponse.of(sliced, query.size, item => `${item.vaultId}${item.blockHeight}`)
+          }
+          return await loop(limit + 30) // loop with larger size call until get the vaultId
+        }
+        return await loop()
+      }
```
Comment (canonbrother) [inline src/module.api/loan.vault.service.ts:125]: why?? because the filtered by `vaultId` and `index` don't work

Comment (fuxingloh) [reply src/module.api/loan.vault.service.ts:125]: Oh damn, we can't do `/vaults/:vaultId/auctions/:batch/history`?

Comment (fuxingloh) [reply src/module.api/loan.vault.service.ts:125]: On Front End we need `Vault -> Auction -> Batch -> History`, means we need to filter by `vaultId` then `batchId` and get many history.

Comment (canonbrother) [reply src/module.api/loan.vault.service.ts:125]: yes the only method now is doing filter on our end
grab full list -> find the target vaultId batch -> return

Comment (fuxingloh) [reply src/module.api/loan.vault.service.ts:125]: Yeah, damn you are right I just checked the LSM Key on ain.

Comment (fuxingloh) [reply src/module.api/loan.vault.service.ts:125]: Could you index `AuctionBatchHistory` on the whale side, perhaps prioritize this task. The front end needs a performant endpoint, I don't think we can use the current implementation.

**Pull request closed without merging**